### ISSUE TITLE
Job tracking fails with `TypeError: '>' not supported between instances of 'float' and 'NoneType'`

### ISSUE THREAD

**Issue by ajs6f** [galaxyproject/galaxy#13298]
**Describe the bug**
A job that actually fails to execute on the cluster to which it is submitted is not reported as such. Instead, it is reported as "currently running" and logs show a cycle around a `TypeError` in `univa.py` (as shown in log excerpts below). What is particularly weird is that the Galaxy log clearly shows the correct (failed) exit code from the job and the correct reason it failed.

**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: (check <galaxy_url>/api/version if you don't know)
release_21.09
Commit: (run `git rev-parse HEAD` if you run this Galaxy server)
81f27cb5878b0e353353f0eefa8a125fe1b5314e
**To Reproduce**
1. Start job that fails on cluster
2. Job is never reported as finishing or failing.
3. Logs show endless repetition as shown below.

**Expected behavior**
Galaxy reports that job has indeed failed.

**Additional context**
Log extract:
```
galaxy.jobs.runners.univa ERROR 2022-02-02 15:22:30,323 [pN:main.web.1,p:123445,w:1,m:0,tN:UnivaJobRunner.monitor_thread] DRMAAUniva: job 2750708 was killed by signal 9
galaxy.jobs.runners.univa ERROR 2022-02-02 15:22:30,323 [pN:main.web.1,p:123445,w:1,m:0,tN:UnivaJobRunner.monitor_thread] DRMAAUniva: job 2750708 failed with failure 46 : execd enforced h_vmem limit
galaxy.jobs.runners ERROR 2022-02-02 15:22:30,324 [pN:main.web.1,p:123445,w:1,m:0,tN:UnivaJobRunner.monitor_thread] Unhandled exception checking active jobs
Traceback (most recent call last):
  File "lib/galaxy/jobs/runners/__init__.py", line 707, in monitor
    self.check_watched_items()
  File "lib/galaxy/jobs/runners/drmaa.py", line 338, in check_watched_items
    if self._complete_terminal_job(ajs, drmaa_state=state) is not None:
  File "lib/galaxy/jobs/runners/univa.py", line 97, in _complete_terminal_job
    elif ("signal" in extinfo and extinfo["signal"] == "SIGKILL") and time_wasted > time_granted:
TypeError: '>' not supported between instances of 'float' and 'NoneType'
10.254.127.6 - - [02/Feb/2022:15:22:30 -0400] "GET /api/histories/f597429621d6eb2b/contents?details=33b43b4e7093c91f&order=hid&v=dev&q=update_time-ge&q=deleted&q=purged&qv=2022-02-02T20%3A22%3A25.000Z&qv=False&qv=False HTTP/1.1" 200 - "https://hydra-5.si.edu/datasets/33b43b4e7093c91f/details" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36"
[pid: 123445|app: 0|req: 176/176] 10.254.127.6 () {86 vars in 2134 bytes} [Wed Feb  2 15:22:30 2022] GET /api/histories/f597429621d6eb2b/contents?details=33b43b4e7093c91f&order=hid&v=dev&q=update_time-ge&q=deleted&q=purged&qv=2022-02-02T20%3A22%3A25.000Z&qv=False&qv=False => generated 2 bytes in 35 msecs (HTTP/1.1 200) 3 headers in 139 bytes (2 switches on core 1)
 galaxy.jobs.runners.drmaa DEBUG 2022-02-02 15:22:31,398 [pN:main.web.1,p:123445,w:1,m:0,tN:UnivaJobRunner.monitor_thread] (5/2750708) state change: job finished normally
lib/galaxy/jobs/handler.py:359: SAWarning: Coercing Subquery object into a select() for use in IN(); please pass a select() construct explicitly
  ~model.Job.table.c.id.in_(hda_not_ready),
lib/galaxy/jobs/handler.py:360: SAWarning: Coercing Subquery object into a select() for use in IN(); please pass a select() construct explicitly
  ~model.Job.table.c.id.in_(ldda_not_ready))
10.254.127.6 - - [02/Feb/2022:15:22:35 -0400] "GET /api/histories/f597429621d6eb2b/contents?details=33b43b4e7093c91f&order=hid&v=dev&q=update_time-ge&q=deleted&q=purged&qv=2022-02-02T20%3A22%3A30.000Z&qv=False&qv=False HTTP/1.1" 200 - "https://hydra-5.si.edu/datasets/33b43b4e7093c91f/details" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36"
[pid: 123445|app: 0|req: 177/177] 10.254.127.6 () {86 vars in 2134 bytes} [Wed Feb  2 15:22:35 2022] GET /api/histories/f597429621d6eb2b/contents?details=33b43b4e7093c91f&order=hid&v=dev&q=update_time-ge&q=deleted&q=purged&qv=2022-02-02T20%3A22%3A30.000Z&qv=False&qv=False => generated 2 bytes in 47 msecs (HTTP/1.1 200) 3 headers in 139 bytes (2 switches on core 2)
galaxy.jobs.runners.univa ERROR 2022-02-02 15:22:36,561 [pN:main.web.1,p:123445,w:1,m:0,tN:UnivaJobRunner.monitor_thread] DRMAAUniva: job 2750708 was killed by signal 9
galaxy.jobs.runners.univa ERROR 2022-02-02 15:22:36,562 [pN:main.web.1,p:123445,w:1,m:0,tN:UnivaJobRunner.monitor_thread] DRMAAUniva: job 2750708 failed with failure 46 : execd enforced h_vmem limit
galaxy.jobs.runners ERROR 2022-02-02 15:22:36,562 [pN:main.web.1,p:123445,w:1,m:0,tN:UnivaJobRunner.monitor_thread] Unhandled exception checking active jobs
Traceback (most recent call last):
  File "lib/galaxy/jobs/runners/__init__.py", line 707, in monitor
    self.check_watched_items()
  File "lib/galaxy/jobs/runners/drmaa.py", line 338, in check_watched_items
    if self._complete_terminal_job(ajs, drmaa_state=state) is not None:
  File "lib/galaxy/jobs/runners/univa.py", line 97, in _complete_terminal_job
    elif ("signal" in extinfo and extinfo["signal"] == "SIGKILL") and time_wasted > time_granted:
TypeError: '>' not supported between instances of 'float' and 'NoneType'

```

**Comment by ajs6f** [comment]
I have removed our ` job_metrics_conf.xml` because it was changed recently and I thought it might be part of this problem. The problem did indeed go away, but I did see several occurances in the log of
```
alaxy.jobs.runners ERROR 2022-02-02 16:22:53,652 [pN:main.web.1,p:138199,w:1,m:0,tN:UnivaJobRunner.monitor_thread] Unhandled exception checking active jobs
Traceback (most recent call last):
  File "lib/galaxy/jobs/runners/__init__.py", line 707, in monitor
    self.check_watched_items()
  File "lib/galaxy/jobs/runners/drmaa.py", line 338, in check_watched_items
    if self._complete_terminal_job(ajs, drmaa_state=state) is not None:
  File "lib/galaxy/jobs/runners/univa.py", line 75, in _complete_terminal_job
    state = self._get_drmaa_state(ajs.job_id, self.ds, True, extinfo)
  File "lib/galaxy/jobs/runners/univa.py", line 454, in _get_drmaa_state
    wstate = self._get_drmaa_state_wait(job_id, ds, extinfo)
  File "lib/galaxy/jobs/runners/univa.py", line 401, in _get_drmaa_state_wait
    extinfo["slots"] = float(rv.resourceUsage['slots'])
KeyError: 'slots'
```
which already [seems to have been reported](https://github.com/galaxyproject/galaxy/issues/9987). I'm not sure whether this ticket is about a problem with the Univa connector or a problem with the way it uses job metrics configuration...

**Comment by bernt-matthias** [comment]
Can you add a log statement printing rv.resourceUsage in lib/galaxy/jobs/runners/univa.py?

Would be interesting for a single core job and a multi core job.

I can't test myself anymore since we moved to slurm.

**Comment by ajs6f** [comment]
Thanks, can you please give me a more precise suggestion? Where in `univa.py` would be a good place to insert this? I'm not a Python programmer and am very much "flying by instrument".

I cannot yet try this on different single- vs. multi-core jobs because we do not yet have reliable operation at all (many segfaults out of Python, perhaps half our tasks complete at all). If (hopefully when) we can reliably execute jobs, I will be happy to set up a test of this kind.

**Comment by bernt-matthias** [comment]
This would be about here https://github.com/galaxyproject/galaxy/blob/931d195c8aa6458fd7cffe8d47abebb279f3ee21/lib/galaxy/jobs/runners/univa.py#L76

Can you also post the content of your `job_conf.xml`?

**Comment by ajs6f** [comment]
Thanks, I will try to get some logging going. I'm currently engaged on your Gitter site trying to get through the aforementioned segfaults, so it may be a little while. In the meantime, I'll include our `job_conf.xml` below.

```
<?xml version="1.0"?>
<!-- A sample job config that describes all available options -->
<job_conf>
    <plugins workers="4">
        <!-- "workers" is the number of threads for the runner's work queue.
             The default from <plugins> is used if not defined for a <plugin>.
             For all asynchronous runners (i.e. everything other than
             LocalJobRunner), this is the number of threads available for
             starting and finishing jobs. For the LocalJobRunner, this is the
             number of concurrent jobs that Galaxy will run.
          -->
        <plugin id="local" type="runner" load="galaxy.jobs.runners.local:LocalJobRunner"/>
        <plugin id="uge" type="runner" load="galaxy.jobs.runners.univa:UnivaJobRunner">
            <!-- Override the $DRMAA_LIBRARY_PATH environment variable -->
            <param id="drmaa_library_path">/cm/shared/apps/uge/current/drmaa/lib/lx-amd64/libdrmaa.so</param>
        </plugin>
    </plugins>


    <handlers>
        <!-- Job handler processes - for a full discussion of job handlers, see the documentation at:
             https://docs.galaxyproject.org/en/latest/admin/scaling.html

             For documentation on handler assignment methods, see the documentation under:
             https://docs.galaxyproject.org/en/latest/admin/scaling.html#job-handler-assignment-methods

             The <handlers> container tag takes four optional attributes:

               <handlers assign_with="method" max_grab="count" ready_window_size="100" default="id_or_tag"/>

               - `assign_with` - How jobs should be assigned to handlers. The value can be a single method or a
                 comma-separated list that will be tried in order. The default depends on whether any handlers and a job
                 handler "pool" (such as uWSGI mules in a `job-handlers` farm) are configured. Valid methods are
                 explained in detail in the documentation referenced above and have the values:

                   - `mem-self` - In-memory Self Assignment
                   - `db-self`- Database Self Assignment
                   - `db-preassign` - Database Preassignment
                   - `uwsgi-mule-message` - uWSGI Mule Messaging
                   - `db-transaction-isolation` - Database Transaction Isolation
                   - `db-skip-locked` - Database SKIP LOCKED

               - `max_grab` - Limit the number of jobs that a handler will self-assign per jobs-ready-to-run check loop
                 iteration. This only applies for methods that cause handlers to self-assign multiple jobs at once
                 (db-skip-locked, db-transaction-isolation) and the value is an integer > 0. Default is to grab as many
                 jobs ready to run as possible.

               - `ready_window_size` - Handlers query for and dispatch jobs ready to run in a loop. If the number of
                 jobs in the `new` state grows very large but they cannot be dispatched due to user concurrency limits,
                 checking all of these jobs on every iteration can drastically slow down job throughput as each pending
                 job is checked and then deferred due to limits. This option limits the number of jobs per user are
                 returned by the query on each iteration. By default it is set to 100. Setting this lower can improve
                 throughput for other users when one user submits thousands of independent jobs.

                 Be aware that anonymous users are treated as a single user by this algorithm.

               - `default` - An ID or tag of the handler(s) that should handle any jobs not assigned to a specific
                 handler (which is probably most of them). If unset, the default is any untagged handlers plus any
                 handlers in the `job-handlers` (no tag) pool.
        -->
        <!-- Explicitly defined job handlers - the id should match the handler process's `server_name`. For webless
             handlers, this is the value of the `server-name` argument to `galaxy-main`.
         -->
<!--        <handler id="uge_handler">
            <plugin id="uge"/>
        </handler> -->
    </handlers> 
    <destinations default="uge-destination">
        <!-- Destinations define details about remote resources and how jobs
             should be executed on those remote resources.
         -->
        <destination id="uge-destination" runner="uge">
            <env id="SGE_ROOT">/cm/shared/apps/uge/current</env>
        </destination>

        
        <destination id="local" runner="local"/>
        <destination id="multicore_local" runner="local">
          <param id="local_slots">4</param> <!-- Specify GALAXY_SLOTS for local jobs. -->
          <!-- Warning: Local slot count doesn't tie up additional worker threads, to prevent over
               allocating machine define a second local runner with different name and fewer workers
               to run this destination. -->
          <param id="embed_metadata_in_job">True</param>
          <!-- Above parameter will be default (with no option to set
               to False) in an upcoming release of Galaxy, but you can
               try it early - it will slightly speed up local jobs by
               embedding metadata calculation in job script itself.
          -->
          <param id="metadata_strategy">directory</param>
          <!-- If set, overwrites metadata_strategy (configurable in galaxy.yml).
               Valid values are `directory` (default), `extended` and `legacy`.
               Setting metadata_strategy to `extended` requires that all object stores
               that can be written to store datasets by uuid. In `extended` mode jobs
               will decide if a tool run failed, the object stores configuration
               is serialized and made available to the job and is used for writing output
               datasets to the object store as part of the job and dynamic output discovery
               (e.g. discovered datasets <discover_datasets>, unpopulated collections, etc)
               happens as part of the job.
          -->
          <job_metrics />
          <!-- Above element demonstrates embedded job metrics definition - see
               job_metrics_conf.xml.sample for full documentation on possible nested
               elements. This empty block will simply disable job metrics for the
               corresponding destination. -->
        </destination>
    </destinations>
    <limits>
        <!-- Certain limits can be defined. The 'concurrent_jobs' limits all
             control the number of jobs that can be "active" at a time, that
             is, dispatched to a runner and in the 'queued' or 'running'
             states.

             A race condition exists that will allow destination_* concurrency
             limits to be surpassed when multiple handlers are allowed to
             handle jobs for the same destination. To prevent this, assign all
             jobs for a specific destination to a single handler.
        -->
        <!-- registered_user_concurrent_jobs:
                Limit on the number of jobs a user with a registered Galaxy
                account can have active across all destinations.
        -->
        <limit type="registered_user_concurrent_jobs">2</limit>
        <!-- anonymous_user_concurrent_jobs:
                Likewise, but for unregistered/anonymous users.
        -->
        <limit type="anonymous_user_concurrent_jobs">1</limit>
        <!-- destination_user_concurrent_jobs:
                The number of jobs a user can have active in the specified
                destination, or across all destinations identified by the
                specified tag. (formerly: concurrent_jobs)
        -->
        <limit type="destination_user_concurrent_jobs" id="local">1</limit>
        <limit type="destination_user_concurrent_jobs" tag="mycluster">2</limit>
        <limit type="destination_user_concurrent_jobs" tag="longjobs">1</limit>
        <!-- destination_total_concurrent_jobs:
                The number of jobs that can be active in the specified
                destination (or across all destinations identified by the
                specified tag) by any/all users.
        -->
        <limit type="destination_total_concurrent_jobs" id="local">16</limit>
        <limit type="destination_total_concurrent_jobs" tag="longjobs">100</limit>
        <!-- walltime:
                Amount of time a job can run (in any destination) before it
                will be terminated by Galaxy.
         -->
        <limit type="walltime">24:00:00</limit>
        <!-- total_walltime:
                Total walltime that jobs may not exceed during a set period.
                If total walltime of finished jobs exceeds this value, any
                new jobs are paused.  `window` is a number in days,
                representing the period.
         -->
        <limit type="total_walltime" window="30">24:00:00</limit>
        <!-- output_size:
                Size that any defined tool output can grow to before the job
                will be terminated. This does not include temporary files
                created by the job. Format is flexible, e.g.:
                '10GB' = '10g' = '10240 Mb' = '10737418240'
        -->
        <limit type="output_size">10GB</limit>
    </limits>
    <macros>
        <xml name="foohost_destination" tokens="id,walltime,ncpus">
            <destination id="@ID@" runner="cli">
                <param id="shell_plugin">SecureShell</param>
                <param id="job_plugin">Torque</param>
                <param id="shell_username">galaxy</param>
                <param id="shell_hostname">foohost_destination.example.org</param>
                <param id="job_Resource_List">walltime=@WALLTIME@,ncpus=@NCPUS@</param>
            </destination>
        </xml>
    </macros>
</job_conf>
```

**Comment by bernt-matthias** [comment]
I see. So you don't use `nativeSpecification` in the UGE destination. 

```
 <param id="nativeSpecification">-l h_rt=24:00:00,h_vmem=8G</param>
```

which might be fine if you do not need to specify it for your cluster. Unfortunately the univa runner in Galaxy currently assumes that you have nativeSpecification. 

I guess it's easy to fix, but I'm not sure how soon I can find time for it. 

Maybe you could try what happens if you add a mock `nativeSpecification` to your destination .. maybe we can trick the system until we have a proper fix.

**Comment by bernt-matthias** [comment]
@ajs6f can you test this: https://github.com/galaxyproject/galaxy/pull/13343

**Comment by ajs6f** [comment]
Yes, I will be very happy to, once I get these segfaults squared away. Thank you for your attention to this issue!

**Comment by bernt-matthias** [comment]
> Yes, I will be very happy to, once I get these segfaults squared away.

Are these segfaults of the python running Galaxy, or segfaults happening in a job? 

If the later, you may use another tool that runs stable.

**Comment by ajs6f** [comment]
In a job, and we cannot run another tool stably because we cannot import any data stably. Every once in a while there is a seemingly random success, but trying to use tools on a successful import generally segfaults out. We haven't gotten that far, so to speak! :)

**Comment by nsoranzo** [comment]
Fixed by #13343.
